presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/KDBTree.java

```diff
@@ -0,0 +1,346 @@
+    {
+        if (node.extent.disjoint(envelope)) {
+            return;
+        }
+
+        if (node.leafId.isPresent()) {
+            leafNodes.put(node.leafId.getAsInt(), node.extent);
+        }
+        else {
+            addIntersectingLeafNodes(envelope, node.left.get(), leafNodes);
+            addIntersectingLeafNodes(envelope, node.right.get(), leafNodes);
+        }
+    }
+
+    private static final Comparator<Rectangle> BY_X = (first, second) -> ComparisonChain.start()
+            .compare(first.getXMin(), second.getXMin())
+            .compare(first.getYMin(), second.getYMin())
+            .result();
+
+    private static final Comparator<Rectangle> BY_Y = (first, second) -> ComparisonChain.start()
+            .compare(first.getYMin(), second.getYMin())
+            .compare(first.getXMin(), second.getXMin())
+            .result();
+
+    private static final class LeafIdAllocator
+    {
+        private int nextId;
+
+        public int next()
+        {
+            return nextId++;
+        }
+    }
+
+    public static KDBTree buildKDBTree(int maxItemsPerNode, int maxLevels, Rectangle extent, List<Rectangle> items)
+    {
+        checkArgument(maxItemsPerNode > 0, "maxItemsPerNode must be > 0");
+        checkArgument(maxLevels > 0, "maxLevels must be > 0");
+        requireNonNull(extent, "extent is null");
+        requireNonNull(items, "items is null");
+        return new KDBTree(buildKDBTreeNode(maxItemsPerNode, maxLevels, 0, extent, items, new LeafIdAllocator()));
+    }
+
+    private static Node buildKDBTreeNode(int maxItemsPerNode, int maxLevels, int level, Rectangle extent, List<Rectangle> items, LeafIdAllocator leafIdAllocator)
+    {
+        checkArgument(maxItemsPerNode > 0, "maxItemsPerNode must be > 0");
+        checkArgument(maxLevels > 0, "maxLevels must be > 0");
+        checkArgument(level >= 0, "level must be >= 0");
+        requireNonNull(extent, "extent is null");
+        requireNonNull(items, "items is null");
+
+        if (items.size() <= maxItemsPerNode || level >= maxLevels) {
+            return new Node(extent, leafIdAllocator.next());
+        }
+
+        // Split over longer side
+        boolean splitVertically = extent.getWidth() >= extent.getHeight();
+        Optional<SplitResult> splitResult = trySplit(splitVertically, maxItemsPerNode, maxLevels, level, extent, items, leafIdAllocator);
+        if (!splitResult.isPresent()) {
+            // Try spitting by the other side
+            splitResult = trySplit(!splitVertically, maxItemsPerNode, maxLevels, level, extent, items, leafIdAllocator);
+        }
+
+        if (!splitResult.isPresent()) {
+            return new Node(extent, leafIdAllocator.next());
+        }
+
+        return new Node(extent, splitResult.get().getLeft(), splitResult.get().getRight());
+    }
+
+    private static final class SplitResult
```

Comment:
nit: You can make it generic, and reuse in `splitVertically` and `splitHorizontally`